pkg/reconciler/identity/iam/iam_policy.go

```diff
@@ -0,0 +1,276 @@
+/*
+Copyright 2020 Google LLC.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package iam
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"cloud.google.com/go/iam"
+	"cloud.google.com/go/iam/admin/apiv1"
+	gclient "github.com/google/knative-gcp/pkg/gclient/iam/admin"
+	iampb "google.golang.org/genproto/googleapis/iam/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type action = int
+
+const (
+	actionAdd action = iota
+	actionRemove
+)
+
+type GServiceAccount string
+type RoleName iam.RoleName
+
+type modificationRequest struct {
+	serviceAccount GServiceAccount
+	role           iam.RoleName
+	member         string
+	action         action
+	respCh         chan error
+}
+
+type roleModification struct {
+	addMembers    sets.String
+	removeMembers sets.String
+}
+
+type batchedModifications struct {
+	roleModifications map[iam.RoleName]*roleModification
+	listeners         []chan<- error
+}
+
+type getPolicyResponse struct {
+	account GServiceAccount
+	policy  *iam.Policy
+	err     error
+}
+
+type setPolicyResponse struct {
+}
+
+// IAMPolicyManager is an interface for making changes to a Google service account's IAM policy.
+type IAMPolicyManager interface {
+	AddIAMPolicyBinding(ctx context.Context, account GServiceAccount, member string, role RoleName) error
+	RemoveIAMPolicyBinding(ctx context.Context, account GServiceAccount, member string, role RoleName) error
+}
+
+var (
+	globalManager       IAMPolicyManager
+	createGlobalManager sync.Once
+)
+
+// DefaultIAMPolicyManager returns a shared global policy manager.
+func DefaultIAMPolicyManager() IAMPolicyManager {
+	createGlobalManager.Do(func() {
+		c, err := admin.NewIamClient(context.Background())
+		if err != nil {
+			panic(err)
+		}
+		m, err := NewIAMPolicyManager(context.Background(), c)
+		if err != nil {
+			panic(err)
+		}
+		globalManager = m
+	})
+	return globalManager
+}
+
+// manager is an IAMPolicyManager which serializes and batches IAM policy changes to a Google
+// Service Account to avoid conflicting changes.
+type manager struct {
+	iam         gclient.IamClient
+	requestCh   chan *modificationRequest
+	pending     map[GServiceAccount]*batchedModifications // a non-nil batch indicates an outstanding request
+	getPolicyCh chan *getPolicyResponse
```

Comment:
It seems you use this to serialize policy updates for an SA since there is one policy associated with an SA? If so, can you add a comment here?

Reply:
I would say that it's the manage process that serializes updates. This channel is used to communicate responses from GetIAMPolicy calls. I've added additional documentation to manage, let me know if anything is still unclear.

Reply:
The updated documentation really helps! Thanks!